### PR TITLE
[DO NOT MERGE JUST YET] PR #90

### LINKED ISSUE #90: PR #89 must be reviewed to handle min and max dates
It seems we also need to change _isDisableDays function to do logic with UTC instead with local TZ.

### PR DESCRIPTION
**Pull request by userquin**
Fixes #90 

I have found a little problem: you only can use min and max with **/** separators (it is possible to bind with ISO format both values, I have had to change my widget to make it work), if these values are provided with ISO format there is no way to make it work (I have spent 2 hours trying to solve the problem but I have not found any solution).

FYI:
I have removed _minDateObj and _maxDateObj to be computed only once (as 2 properties with computed callback), that is, both are calculated when minDate and maxDate changes respectively, so _isDisableDay dont need to recompute both for each date in the calendar.

---
This change is [Reviewable](https://reviewable.io/reviews/motss/app-datepicker/91)

### CODE REVIEW

**Comment by userquin** [comment]
Do not merge this pull request, I have to review it, I think I have the solution but maybe some breaking change may be introduced due how date constructor is used

**Comment by motss** [comment]
@userquin I thought we needed to change all Date to UTC format, so that it won't use the user's system timezone offset. So, as long as we use UTC datetime, we will be safe to compare the `minDate` and `maxDate` with the current date since all Dates are UTC based.

**Comment by userquin** [comment]
Yes, thats the problem, check these tests:

```javascript
new Date('2015-01-01').toISOString();
"2015-01-01T00:00:00.000Z"
new Date('2015-01-01'.split('-')).toISOString();
"2014-12-31T23:00:00.000Z"
new Date('2015/01/01').toISOString();
"2014-12-31T23:00:00.000Z"
new Date(2015,1,1).toISOString();
"2015-01-31T23:00:00.000Z"
new Date(2015,1,1).getTime();
1422745200000
new Date('2015/01/01').getTime();
1420066800000
new Date('2015-01-01').getTime();
1420070400000
```

The problem is that min, max and disabled days can be any dateString that new Date can parse. In my new branch I force to use ISO format for min,max and disabled days, so I can then use client TZ, just because new Date("yyyy-mm-dd") always is the same, no matter UTC or client TZ.

I have it working on my branch, I'm updating tests to pass all, I have changes many things wrong in the tests.

Also I have updated the locale calculation to be OK.

**Comment by motss** [comment]
@userquin 

Hmm... Some of it are not recommended way to do it as the way of how the web platform deals with Date object is yet to standardized as not all web platforms will treat it the same way as you expected and apparently this is still a serious issue in most modern web browsers these days. That's why there is a proposal aims to solve this kind of weird Date object for once and for all. 😃 

So for me, as you might find out in the code, sometimes I do a lot of things like keep using `getFullYear`, `getMonth`, `getDate` to form a new Date object. That's my way to ensure compatibility and correctness when dealing with the nasty Date object in Javascript.

This is the way I talked you about when dealing with `minDate` and `maxDate`, worth taking a look:

```js
/* Always default to UTC time */
/* Let say minDate has a value of '2017-01-01' */
const minDate = new Date('2017-01-01'); /* Surprisingly this will automatically in UTC offset in only modern browsers but not '2017/01/01' */
// These are equivalent to the above line
// const minDate = new Date(Date.UTC(2017, 0, 1)); // Don't forget month 0 is January

/* To compare two datetime */
// 1. either use shorthand to get the timestamp
const minDateTimestamp = +minDate;
// 2. or use `getTime` function
// const minDateTimestamp = minDate.getTime();

/* Since everything will be in UTC format (assuming current datetime is in UTC format) */
/* You simply do equality comparison */
const now = new Date();
const fullYear = now.getFullYear();
const month = now.getMonth();
const date = now.getDate();
const currentDateTimestamp = +new Date(Date.UTC(fullYear, month, date));
if (minDateTimestamp === currentDateTimestamp) {
 /* Do something */
} else {
 /* Do something else if not matched */
}
```

**Comment by userquin** [comment]
I agree, but the problem is not UTC or client TZ. As I said, the problem is just the way the constructor is invoked, that is, if dateString is a ISO date string or not. As you can read in Date doccumentation from mozilla, it remarks that it is not recommended to create a Date with no ISO format nor using parseDate with no ISO format.

What is not standarized is the use of non ISO date strings in the constructor nor parseDate. You can still use new Date('yyyy-mm-dd') or new Date(yyyy,mm-1,dd), just you must take into account what does or does not each constructor: first one UTC and last in client TZ.

From the original code:
```javascript
var _minDateObj = this._convertDateStringToDateObject(_minDate); ===>UTC or UTC from client TZ<===
var _maxDateObj = this._convertDateStringToDateObject(_maxDate); ===>UTC or UTC from client TZ<===
var _currentDate = new Date(this._activeYear, this._activeMonth, _item); ===>client TZ<===
```

_convertDateStringToDateObject returns new Date(date), so if the user configures a non ISO date string the returned date will be in **UTC from client TZ (UTC +- TZ timezone offset)** while providing a ISO date string the returned date will be in **UTC**.

Following with the original code:
```javascript
// run two different obj differently just in case only one of them
// is defined and still be able to update disabled days.
if (_minDateObj) {
  _isLessThanMinDate = _currentDate.getTime() < _minDateObj.getTime();
}
if (_maxDateObj) {
  _isMoreThanMaxDate = _currentDate.getTime() > _maxDateObj.getTime();
}
```
you are comparing two different things, see my previous results: _currentDate is in client TZ and _minDateObj and/or _maxDateObj will be in UTC or client TZ:
```javascript
new Date(2015,1,1).getTime(); ===>this is how _currentDate is built<===
1422745200000
new Date('2015/01/01').getTime();===>this is how _minDateObj or _maxDateObj is built<===
1420066800000
new Date('2015-01-01').getTime();===>this is how _minDateObj or _maxDateObj is built<===
1420070400000
```

In the other hand, if all dates are in ISO format, I can use Date.UTC(y,m,d) or new Date('yyyy-mm-dd') or new Date('yyyy-mm-dd').split('-') just because now all things are the same and now I can compare them.

**Comment by userquin** [comment]
With original code, my widget does not work just because I provide min and max as ISO strings, so there was a day included or excluded, depending on client TZ

**Comment by userquin** [comment]
Take a look at PR #92

**Comment by motss** [comment]
@userquin By the way, too many PRs are not a good thing. Very difficult for you and me to review. Please do the PR one by one or all in one. Be sure what to solve in each PR.

So our aim is to make everything in *UTC* format and these are the ways to handle that:

1. If __date__ is of *ISO* format, use it directly with no fear - it's in *UTC* format.
2. If __date__ is not of *ISO* format - let's say a date string `2017-01-01` - we use `RegExp` to extract out the parts like `yyyy`, `mm`, and `dd` to reconstruct a proper date string in *ISO* format or make it a `Date` object that's in *UTC* format.

In short, the `date` should not depend on the user's `TZ` at all as the timezone offset will make things work unexpectedly.

**Comment by userquin** [comment]
you can ignore/remove PR #90, the good one is PR #92

**Comment by motss** [comment]
Then how about #91

**Comment by userquin** [comment]
there are only 2 PR, PR #91 (this) and PR #92 The names of the PR are confusing us, sorry for my mistake on the name of this PR. You can ignore/remove this PR.

**Comment by motss** [comment]
Closing...
